libraries/c_sdk/standard/mqtt/src/iot_mqtt_subscription.c

```diff
@@ -318,70 +299,76 @@ IotMqttError_t _IotMqtt_AddSubscriptions( _mqttConnection_t * pMqttConnection,
+            topicMatchParams.topicNameLength = pSubscriptionList[ i ].topicFilterLength;
+            /* Finding the matching subscription if it exists. */
+            matchedIndex = IotMqtt_FindFirstMatch( &( connToContext[ contextIndex ].subscriptionArray ), 0, &topicMatchParams );
 
-            /* The lengths of exactly matching topic filters must match. */
-            IotMqtt_Assert( pNewSubscription->topicFilterLength == pSubscriptionList[ i ].topicFilterLength );
-
-            /* Replace the callback and packet info with the new parameters. */
-            pNewSubscription->callback = pSubscriptionList[ i ].callback;
-            pNewSubscription->packetInfo.identifier = subscribePacketIdentifier;
-            pNewSubscription->packetInfo.order = i;
-        }
-        else
-        {
-            /* Allocate memory for a new subscription. */
-            pNewSubscription = IotMqtt_MallocSubscription( sizeof( _mqttSubscription_t ) +
-                                                           pSubscriptionList[ i ].topicFilterLength );
-
-            if( pNewSubscription == NULL )
+            if( matchedIndex != -1 )
             {
-                status = IOT_MQTT_NO_MEMORY;
-                break;
+                IotMqtt_Assert( connToContext[ contextIndex ].subscriptionArray[ matchedIndex ].topicFilterLength == pSubscriptionList[ i ].topicFilterLength );
+
+                /* Replace the callback and packet info with the new parameters. */
+                connToContext[ contextIndex ].subscriptionArray[ matchedIndex ].callback = pSubscriptionList[ i ].callback;
+                connToContext[ contextIndex ].subscriptionArray[ matchedIndex ].packetInfo.identifier = subscribePacketIdentifier;
+                connToContext[ contextIndex ].subscriptionArray[ matchedIndex ].packetInfo.order = i;
             }
             else
             {
-                /* Clear the new subscription. */
-                ( void ) memset( pNewSubscription,
-                                 0x00,
-                                 sizeof( _mqttSubscription_t ) + pSubscriptionList[ i ].topicFilterLength );
-
-                /* Set the members of the new subscription and add it to the list. */
-                pNewSubscription->packetInfo.identifier = subscribePacketIdentifier;
-                pNewSubscription->packetInfo.order = i;
-                pNewSubscription->callback = pSubscriptionList[ i ].callback;
-                pNewSubscription->topicFilterLength = pSubscriptionList[ i ].topicFilterLength;
-                ( void ) memcpy( pNewSubscription->pTopicFilter,
-                                 pSubscriptionList[ i ].pTopicFilter,
-                                 ( size_t ) ( pSubscriptionList[ i ].topicFilterLength ) );
-
-                IotListDouble_InsertHead( &( pMqttConnection->subscriptionList ),
-                                          &( pNewSubscription->link ) );
+                /* Allocate memory for a new subscription. */
```

Comment:
Is it allocating memory for subscription?

Reply:
Addressed